### PR TITLE
[New] `no-unused-modules`: support dynamic imports 

### LINKED ISSUE #1951: [no-unused-modules] Dynamic import doesn't mark exports as used
Repro:

```ts
// foo.ts
export const foo = "foo"

// index.ts
async function test() {
  const { foo } = await import("./foo")
  const foo2 = await import("./foo").then(mod => mod.foo)
  const mod = await import("./foo") // should mark all non-type exports as used
}
```

My specific use case is:

```tsx
import { lazy } from "react"
const Foo = lazy(() => import("./Foo"))
```

> exported declaration 'foo' not used within other modules

Config:

```json5
{
  "extends": [
    "plugin:import/recommended",
    "plugin:import/typescript",
    // and more ...
  ],
  "rules": {
    "import/no-unused-modules": ["error", { "unusedExports": true }],
  },
}
```

Versions:

```
@typescript-eslint/parser: 4.8.0
eslint: 7.13.0
eslint-plugin-import: 2.22.1
```

### PR DESCRIPTION
**Pull request by maxkomarychev**
All occurences of `import('...')` are treated as namespace imports
(`import * as X from '...'`)

Fixes #1951.

### CODE REVIEW

**Comment by ljharb** [comment]
`const a = await import('a')` is not equivalent to `import a from 'a'`, it's equivalent to getting a promise for `import * as mod from 'a'`.

`await` is largely irrelevant here, i think.

Because `import()` produces a promise, I don't think there's any value in checking that specific imports/exports are used or not, because it's not very reliably to detect that.

**Comment by maxkomarychev** [comment]
hey, thanks for feedback!

> const a = await import('a') is not equivalent to import a from 'a', it's equivalent to getting a promise for import * as mod from 'a'. You might be thinking of const { default: a } = await import('a')

good point, you're right - I should be treating this as namespace :)

> await is largely irrelevant here, i think.
> Because import() produces a promise, I don't think there's any value in checking that specific imports/exports are used or not, because it's not very reliably to detect that.

Not sure I entirely follow here. I was not going to check if any specific name imported from the module has ever been referenced (there are other eslint rules for this matter) but when import is assigned to destructured object I want to mark specific names that were imported, isn't this good for this rule?

**Comment by ljharb** [comment]
Right, that's what I'm saying - it's not going to be safe in enough cases to mark specific names, so any path that's dynamically imported should be treated as potentially using every export.

**Comment by maxkomarychev** [comment]
> Right, that's what I'm saying - it's not going to be safe in enough cases to mark specific names, so any path that's dynamically imported should be treated as potentially using every export.

Even though handling all kinds of expressions would be very complex I believe is possible to handle some finite amount of patterns and it would increase accuracy of the lib, i.e.

|import base|with await|with then|
|-|-|-|
|`import * as a from 'a'` | `const a = await import('a')` | `import('a').then(a =>{})` or `import('a').then(function(a){})`
|`import { a } from 'a' `|`const { a } = await import('a')` | `import('a').then(({a}) =>{})` or `import('a').then(function(({a}){})`

all other calls to `import('a')` will simply be treated as namespace import

is that something you would consider to review?

**Comment by ljharb** [comment]
In your first row, the `a` could be passed elsewhere. I'm not sure how often the second one happens to be useful.

`import()` is also often used behind an abstraction, or in `Promise.all` like `Promise.all([import('a'), import('b')])`, which is then either awaited or `.then`d. Do you propose attempting to handle those cases as well?

**Comment by maxkomarychev** [comment]
alright, this is too much :) I'll treat `import('a')` as namespace no matter what

File: tests/src/rules/no-unused-modules.js [hunk 1]
```diff
@@ -163,6 +173,40 @@ ruleTester.run('no-unused-modules', rule, {
...
+           const d = 40
+           export default d
+           `,
+           parser: require.resolve('babel-eslint'),
+           filename: testFilePath('./no-unused-modules/exports-for-dynamic-js.js')}),
+  ],
+  invalid: [
+  ],
+})
+typescriptRuleTester.run('no-unused-modules', rule, {
```
Comment (maxkomarychev) [inline tests/src/rules/no-unused-modules.js]: this test always passes however corresponding ts file is empty, what am I doing wrong?

**Comment by maxkomarychev** [comment]
@ljharb I almost cleaned it up although I am stuck at typescript files - I put a dedicated test with typescript parser and it is always green - how do I configure it to fail?

File: tests/src/rules/no-unused-modules.js (hunk 1 above)
Comment (ljharb) [reply tests/src/rules/no-unused-modules.js]: i believe the `filename` doesn't have to exist, it's just to set the context for rules

Comment (maxkomarychev) [reply tests/src/rules/no-unused-modules.js]: hmm. how does it check if specific file is imported? I thought filename here is provided specifically to check that this code belongs to specificed file and then track imports of such file. 

Which in case of ts does not work and tests always green.

**Comment by maxkomarychev** [comment]
ok I believe I figured this out, although one last bit that puzzles me - dynamic imports are supported by espree since 6.1.0 (https://github.com/eslint/espree/commit/f5e58cc5e9030793baca3426366b8d7286ef5b89) but tests fail even if `ecmaVersion` is set to 11. so I can only got it working with `babel-eslint` or `@typescript-eslint/parser`. any suggestion how to make espree to handle this? or would it instead be acceptable to just specify this in docs for specific rule to use custom parser?

**Comment by ljharb** [comment]
It wasn't stage 4 until ES2020, so eslint wouldn't support it until that ecmaVersion.

If it's still not working with the right ecmaVersion, it might not work in eslint until v7; for now, it's fine just to support those two parsers.

**Comment by maxkomarychev** [comment]
should I modify readme with this?
otherwise - ready for review 🎆

**Comment by maxkomarychev** [comment]
it fails tests on eslint 2 and 3, what's the best way to deal with these?

**Comment by maxkomarychev** [comment]
bump?

File: src/ExportMap.js [hunk 2]
```diff
@@ -344,14 +345,43 @@ ExportMap.parse = function (path, content, context) {
...
-    log('parse error:', path, err)
     m.errors.push(err)
     return m // can't continue
   }
 
-  if (!unambiguous.isModule(ast)) return null
+  let hasDynamicImports = false
+
+  visit(ast, visitorKeys, {
+    CallExpression(node) {
```
Comment (Hypnosphi) [inline src/ExportMap.js:357]: It's `ImportExpression` in the latest `@babel/parser` which is a dependency of `babel-eslint`
https://github.com/babel/babel/pull/10828

File: src/ExportMap.js [hunk 3]
```diff
@@ -344,14 +345,43 @@ ExportMap.parse = function (path, content, context) {
...
+      if (node.callee.type === 'Import') {
+        hasDynamicImports = true
+        const firstArgument = node.arguments[0]
```
Comment (Hypnosphi) [inline src/ExportMap.js:360]:
```suggestion
    ImportExpression(node) {
      hasDynamicImports = true
      const source = node.source
        
      // replace `firstArgument` with `source` below
```

Comment (ljharb) [reply src/ExportMap.js:360]: We'd need to keep both; `ImportExpression` is the newer form, but older parsers still have a `CallExpression` of callee type Import.

**Comment by Hypnosphi** [comment]
@maxkomarychev looks like you also have to update https://github.com/benmosher/eslint-plugin-import/blob/master/src/rules/no-unused-modules.js#L610

**Comment by maxkomarychev** [comment]
@Hypnosphi can you elaborate please? this pr puts dynamic imports with as they are namespace imports, given that all other code is using `ExportMap` then it should be transparent to them with no special handling, unless I'm missing something. 

Thanks.

**Comment by Hypnosphi** [comment]
`updateImportUsage` doesn't use `ExportMap`, it does its own iteration instead

**Comment by ljharb** [comment]
@maxkomarychev there's still tests failing, the PR needs a rebase, plus https://github.com/benmosher/eslint-plugin-import/pull/1660#issuecomment-631161376. are you planning to update the PR?

**Comment by maxkomarychev** [comment]
@ljharb yes. although I have a question: I see tests have been failing in lower versions of eslint, do lower versions of eslint support newer syntaxes or maybe you can point to a place where I can see how old versions should be handled in rules? thanks.

**Comment by ljharb** [comment]
It's entirely likely that `import()` won't work in eslint < 6; you can use an approach like this: https://github.com/benmosher/eslint-plugin-import/blob/0547c7efa0a1becd2cdf35ae03f6fe74472398d9/tests/src/cli.js#L32

File: utils/visit.js [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+  for (const fieldName of childFields) {
+    const field = node[fieldName]
+    if (Array.isArray(field)) {
+      for (const item of field) {
+        visit(item, keys, visitorSpec)
+      }
+    } else {
+      visit(field, keys, visitorSpec)
+    }
+  }
```
Comment (ljharb) [inline utils/visit.js:25]:
```suggestion
  childFields.forEach((fieldName) => {
    [].concat(node[fieldName]).forEach((item) => {
      visit(item, keys, visitorSpec)
    })
  })
```

File: utils/visit.js [hunk 5]
```diff
@@ -0,0 +1,29 @@
+'use strict'
+exports.__esModule = true
```
Comment (ljharb) [inline utils/visit.js:2]:
```suggestion
```
this should only ever be added by babel.

File: utils/visit.js [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+
+exports.default = function visit(node, keys, visitorSpec) {
```
Comment (ljharb) [inline utils/visit.js:4]:
```suggestion
module.exports = function visit(node, keys, visitorSpec) {
```

File: utils/parse.js [hunk 7]
```diff
@@ -55,16 +84,25 @@ exports.default = function parse(path, content, context) {
       console.warn(
         '`parseForESLint` from parser `' +
           parserPath +
-          '` is invalid and will just be ignored'
+          '` is invalid and will just be ignored',
+        path
```
Comment (ljharb) [inline utils/parse.js:88]:
```suggestion
          '` is invalid and will just be ignored'
```
if the path is needed, let's put it in the string

File: utils/parse.js [hunk 8]
```diff
@@ -3,9 +3,33 @@ exports.__esModule = true
...
 const log = require('debug')('eslint-plugin-import:parse')
 
+function getBabelVisitorKeys(parserPath) {
+  const hypotheticalLocation = parserPath.replace('index.js', 'visitor-keys.js')
+  if (fs.existsSync(hypotheticalLocation)) {
+    const keys = moduleRequire(parserPath.replace('index.js', 'visitor-keys.js'))
+    return keys
+  } else {
+    return null
+  }
```
Comment (ljharb) [inline utils/parse.js:17]:
```suggestion
  }
  return null
```

File: utils/parse.js [hunk 9]
```diff
@@ -3,9 +3,33 @@ exports.__esModule = true
...
 const moduleRequire = require('./module-require').default
 const extname = require('path').extname
+const fs = require('fs')
 
 const log = require('debug')('eslint-plugin-import:parse')
 
+function getBabelVisitorKeys(parserPath) {
+  const hypotheticalLocation = parserPath.replace('index.js', 'visitor-keys.js')
+  if (fs.existsSync(hypotheticalLocation)) {
+    const keys = moduleRequire(parserPath.replace('index.js', 'visitor-keys.js'))
```
Comment (ljharb) [inline utils/parse.js:13]:
```suggestion
    const keys = moduleRequire(hypotheticalLocation)
```

File: utils/parse.js [hunk 10]
```diff
@@ -3,9 +3,33 @@ exports.__esModule = true
...
+    return null
+  }
+}
+
+function keysFromParser(parserPath, parserInstance, parsedResult) {
+  if (/.*espree.*/.test(parserPath)) {
+    return parserInstance.VisitorKeys
+  } else if (/.*babel-eslint.*/.test(parserPath)) {
+    return getBabelVisitorKeys(parserPath)
+  } else if (/.*@typescript-eslint\/parser/.test(parserPath)) {
```
Comment (ljharb) [inline utils/parse.js:25]:
```suggestion
  }
  if (/.*babel-eslint.*/.test(parserPath)) {
    return getBabelVisitorKeys(parserPath)
  }
  if (/.*@typescript-eslint\/parser/.test(parserPath)) {
```

**Comment by ljharb** [comment]
@maxkomarychev we need to keep both node names - we have to retain compatibility with old version. See https://github.com/benmosher/eslint-plugin-import/pull/1660#discussion_r427657128

**Comment by Hypnosphi** [comment]
@maxkomarychev do you need any help with that? If you're not interested in it anymore, I can open a new PR

**Comment by ljharb** [comment]
@Hypnosphi please don't open a new PR to this repo - either make a PR to the OP's repo (and ping me in it) or comment here with a link to a branch, and i'll update this PR. Thanks!

**Comment by maxkomarychev** [comment]
Hey, I don't have capacity to push this to completion in near future.
I'm sorry.

@Hypnosphi feel free to submit your own version.

**Comment by Hypnosphi** [comment]
@ljharb https://github.com/Hypnosphi/eslint-plugin-import/tree/no-unused-modules-dynamic-imports

**Comment by aladdin-add** [comment]
any news?

**Comment by ljharb** [comment]
@Hypnosphi unfortunately due to the intervening time, there are now merge conflicts between your branch and master. Any chance you could rebase your branch?

**Comment by Hypnosphi** [comment]
I will, but I have now a concern about this implementation though. We heavily [rely](https://github.com/benmosher/eslint-plugin-import/blob/c10e2172c533fc29f7404e6cd4cc628ac3df5f8e/utils/parse.js#L10-L17) on internal file structure of parser packages which can change in minor releases. It actually changed in some of the `@babel/eslint-parser` releases. I figured out a [workaround](https://github.com/Hypnosphi/eslint-plugin-import/commit/513960d9f559fc98bc15a8215e4a63814e6f63d7), but maybe there's a more generic way to find all dynamic imports?

**Comment by Hypnosphi** [comment]
Done

**Comment by ljharb** [comment]
@Hypnosphi arguably anything exposed is part of the api, and can't be moved within a major without violating semver. If the parser refuses to comply with that, then the only other alternative is to convince them to provide the API we need as something that's officially supported.

**Comment by aladdin-add** [comment]
generally, it's not encouraged to do so. But for `babel-eslint`, I think it's fine to do so (`babel-eslint` is no longer maintained, in favor of the new `@babel/eslint-parser`).

**Comment by ljharb** [comment]
lol properties with "_" in front are not encouraged to be relied upon either, but they're public and part of the API anyways ¯\\\_(ツ)\_/¯ reachable is public.

i agree that for babel-eslint the risk is effectively zero; i'm comfortable relying on that.

**Comment by aladdin-add** [comment]
@Hypnosphi Did you forget to push the latest changes? it does not work well on my computer.

1. lint error, I made a PR: https://github.com/Hypnosphi/eslint-plugin-import/pull/1
2. after fixed 1, I got another error when run `npm test`:
```bash
TypeError: Error while loading rule 'no-unused-modules': Cannot read property 'forEach' of undefined
Occurred while linting /Users/weiran/repo/dian/eslint-plugin-import/tests/files/no-unused-modules/typescript/file-ts-e-unused.ts
      at src/rules/no-unused-modules.js:240:28
      at Map.forEach (<anonymous>)
      at src/rules/no-unused-modules.js:235:23
      at Set.forEach (<anonymous>)
      at prepareImportsAndExports (src/rules/no-unused-modules.js:184:12)
      at doPreparation (src/rules/no-unused-modules.js:333:3)
      at create (src/rules/no-unused-modules.js:473:7)
      at Object.create (node_modules/eslint/lib/rule-tester/rule-tester.js:455:73)
      at createRuleListeners (node_modules/eslint/lib/linter/linter.js:761:21)
      at node_modules/eslint/lib/linter/linter.js:931:31
      at Array.forEach (<anonymous>)
      at runRules (node_modules/eslint/lib/linter/linter.js:876:34)
      at Linter._verifyWithoutProcessors (node_modules/eslint/lib/linter/linter.js:1175:31)
      at Linter.verify (node_modules/eslint/lib/linter/linter.js:1238:21)
      at runRuleForItem (node_modules/eslint/lib/rule-tester/rule-tester.js:556:37)
      at testInvalidTemplate (node_modules/eslint/lib/rule-tester/rule-tester.js:654:28)
      at Context.<anonymous> (node_modules/eslint/lib/rule-tester/rule-tester.js:893:25)
      at processImmediate (node:internal/timers:464:21)
```

hope it can be landed soon! :)

**Comment by Hypnosphi** [comment]
@aladdin-add good catch, fixed

**Comment by aladdin-add** [comment]
@Hypnosphi I've tested it on a large repo, and got an error:

```bash
➜  honor git:(4d146d797f) ✗ npx eslint /Users/weiran/repo/dian/honor/src/common/components-new/navigator/index.tsx

Oops! Something went wrong! :(

ESLint: 7.27.0

TypeError: Error while loading rule 'import/no-unused-modules': Cannot read property 'get' of undefined
Occurred while linting /Users/weiran/repo/dian/honor/src/common/components-new/navigator/index.tsx
    at /Users/weiran/repo/dian/honor/node_modules/eslint-plugin-import/lib/rules/no-unused-modules.js:266:44
    at Set.forEach (<anonymous>)
    at /Users/weiran/repo/dian/honor/node_modules/eslint-plugin-import/lib/rules/no-unused-modules.js:264:11
    at Map.forEach (<anonymous>)
    at prepareImportsAndExports (/Users/weiran/repo/dian/honor/node_modules/eslint-plugin-import/lib/rules/no-unused-modules.js:263:13)
    at doPreparation (/Users/weiran/repo/dian/honor/node_modules/eslint-plugin-import/lib/rules/no-unused-modules.js:333:3)
    at Object.create (/Users/weiran/repo/dian/honor/node_modules/eslint-plugin-import/lib/rules/no-unused-modules.js:473:7)
    at createRuleListeners (/Users/weiran/repo/dian/honor/node_modules/eslint/lib/linter/linter.js:761:21)
    at /Users/weiran/repo/dian/honor/node_modules/eslint/lib/linter/linter.js:931:31
    at Array.forEach (<anonymous>)
```

I am using:
* node: v16.1.0
* eslint: v7.27.0
* @typescript-eslint/parser: v4.24.0

the code is something like:
```tsx
import React, { MouseEvent, ReactNode } from 'react'
import './styles.less'

export interface actionItem {
  icon?: string
  node?: ReactNode
  action?: (e: MouseEvent<any>) => void
}

export interface NavigatorProps {
  user: { icon?: string; title?: string }
  onClickLeft?: (e: MouseEvent<any>) => void
  actions?: actionItem[]
}

const Navigator: React.FC<NavigatorProps> = ({
  user = { icon: HEAD_PORTRAIT, title: '' },
  onClickLeft,
  actions,
}) => {
  return (
    <div className="dian-navigator-wraper">
      <div className="dian-navigator-top" />
      <div className="dian-navigator">
        <div
          className="icon-wrapper"
          onClick={e => onClickLeft && onClickLeft(e)}
        >
          <img className="icon" src={user.icon} />
        </div>
        <div className="icon-wrapper search-wrapper">
          <span className="title">{user.title}</span>
        </div>
        {actions?.map((action: actionItem) => (
          <div className="icon-wrapper" onClick={action.action}>
            {action.icon && <img className="icon small" src={action.icon} />}
            {!action.icon && action.node && action.node}
          </div>
        ))}
      </div>
    </div>
  )
}

export default Navigator

```

**Comment by Hypnosphi** [comment]
@aladdin-add just to be clear, did you apply both `eslint-plugin-import` and `eslint-module-utils` from my branch?

**Comment by aladdin-add** [comment]
yes! used yarn's resolutions.

**Comment by Hypnosphi** [comment]
@aladdin-add can you please either send a PR with a failing test or create a reproduction repo? I can't reproduce the issue with the code you provided

**Comment by aladdin-add** [comment]
sure, will create a repo later. :)

**Comment by ljharb** [comment]
@Hypnosphi any update?

**Comment by Hypnosphi** [comment]
@ljharb I tried adding `export * from '@org/package'` to various test files but still wasn't able to reproduce the issue reported by @aladdin-add

**Comment by ljharb** [comment]
@aladdin-add any chance you could come up with a failing test case?

**Comment by aladdin-add** [comment]
I was not able to repro it either - it happened in a large and private repo. /_ \

it seems a very edge case. we can try to fix it when someone else hit the issue?

**Comment by ljharb** [comment]
Sure! This PR then would need a rebase, and then I can land it.

**Comment by aladdin-add** [comment]
friendly ping @ljharb

**Comment by ljharb** [comment]
I'll take a look at this in the next 24 hours or so.

**Comment by ljharb** [comment]
@Hypnosphi perhaps i messed up the rebase, but a bunch of tests are still failing. could you make sure to fetch the latest branch, and then get the tests passing?

**Comment by Hypnosphi** [comment]
Sorry, it's hard to track what's changed here comparing to my branch. I think I'll open a separate PR

**Comment by ljharb** [comment]
@Hypnosphi i really, really wish you had not done that; PRs once opened are *eternally* present in a repo's refs. Now I just have to keep both of these PRs in sync until they're landed.

**Review by ljharb — APPROVED** [review]

**Comment by Evertt** [comment]
Hey, so I'm using this plugin at version 2.26.0, which I believe should include this merged PR already, right? Well I still get a false positive with this.

image

image

So I'm guessing there's still a little bug left in this feature? Can I do anything to help you track down the bug?

**Comment by aladdin-add** [comment]
yes, please file a new issue if you still encounter it.
